---
title: "GxEScanR - nsaid/aspirin variable"
bibliography: /home/rak/Dropbox/library.bib
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    toc_depth: 2
    theme: united
---


```{r setup, echo=F, include=F, cache=T}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(broom) 
library(kableExtra)
library(data.table)
library(rlang)
library(rmeta)
library(qqman)

# rm(list = ls())
# gxe_set <- readRDS("~/git/DATA/FIGI_EpiData_rdata/FIGI_Genotype_Epi_181120_EpiOnly.rds") %>% 
#   filter(drop == 0 & gxe == 1) %>%  # GxE set - N = 102792 as of 12/13/2018
#   mutate(studyname = ifelse(studyname %in% c("HPFS_1", "HPFS_2"), "HPFS_1_2", studyname),
#          studyname = ifelse(studyname %in% c("NHS_1", "NHS_2"), "NHS_1_2", studyname),
#          aspirinyrs_b = ifelse(aspirinyrs > 0, 1, 0),
#          nsaidsyrs_b = ifelse(nsaidsyrs > 0, 1, 0),
#          aspyrs_ref_b = ifelse(aspyrs_ref > 0, 1, 0))

```




# List of NSAID/aspirin variables
- aspirin - aspirin use @ ref time  
- aspirinyrs - aspirin, years taken (any time?)  
- nsaids - non-aspirin NSAIDS use @ ref time  
- nsaidsyrs - NSAIDS, years taken (any time?)  
- aspirin_ever - aspirin, ever used regularly  
- nsaids_ever - NSAIDS, ever used regularly  
- aspirin_nsaids_ever - Aspirin and/or NSAIDS, ever used regularly?  
<br>
- asp_ref - regular aspirin/NSAID use @ ref time (NO if aspirin OR nsaids == no)  
- asp_ref2 - regular aspirin/NSAID use @ ref time (NO if aspirin AND nsaids == no)  
  - [note - asp_ref vs asp_ref2 N = 1771]  
- aspyrs_ref - regular aspirin/NSAID use @ ref time, duration  



# Bdose vs VCF
Ensure that dosages are identical between original imputation VCF files + BDose files  

Steps:  
- GxEScanR nsaid/aspirin (asp_ref) scan  
- N = 91 SNPs with D|G p value < 5e-8  
- Extract those SNPs from both VCF and BDose sources  

```{r dosage_compare, message=F, echo=T, warning=F, tidy=T, cache=T}
load("~/git/Dosage_Extract/vcf_bdose_comparison_objects.RData")

# BDOSE FILE
head(bdose[, c(1:10)])

# VCF FILE
head(vcf[, c(1:10)])

all(vcf == bdose)
```


# GxEScanR run - asp_ref
![](/home/rak/Dropbox/code/FIGI_Results/aspref_sex_age_pc10_studyname/GxEScanR_asp_ref_chiSqG_qq.png)
![](/home/rak/Dropbox/code/FIGI_Results/aspref_sex_age_pc10_studyname/GxEScanR_asp_ref_chiSqG_manhattan.png)

Extract 91 SNPs with p value < 5e-8, run same model using GLM.. 

![](/home/rak/Dropbox/code/FIGI_Results/aspref_sex_age_pc10_studyname/GLM_asp_ref_chiSqG_manhattan.png)

## Compare results tables
(estimates + p values differ between GLM vs GxEScanR)

```{r results_compare, message=F, echo=T, warning=F, tidy=T, cache=T, include = F}
setwd("~/Dropbox/code/FIGI_Results/aspref_sex_age_pc10_studyname/")
results <- do.call(rbind, lapply(list.files(full.names = T, pattern = "results_n"), fread, stringsAsFactors = F))
names(results)

results_G <- results %>%
  separate(SNP, into = c("CHR", "BP"), sep = ":", remove = F) %>% 
  mutate(CHR = as.numeric(CHR), 
         BP = as.numeric(BP), 
         P = pchisq(chiSqG, df = 1, lower.tail = F)) %>% 
  filter(P < 5e-8)


# just for documentation - running GLM on same covariate file
# cov <- readRDS("~/git/GxEScanR_CovFiles/FIGI_GxESet_asp_ref_sex_age_pcs_studyname_72438.rds")
# for(chr in 1:22) {
#   x <- as.data.frame(readRDS(paste0("~/git/Dosage_Extract/GxEScanR_asp_ref_extract_chr", chr, ".rds"))) %>% 
#     rownames_to_column(var = 'vcfid')
#   cov <- inner_join(cov, x, by = 'vcfid')
# }
# 
# glm_func_original <- function(y) glm(outcome ~ y + age_ref_imp + sex + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 +MCCS_1+NFCCR+MECC_2+CCFR_3+Kentucky+CCFR_1+PPS3+CRCGEN+MECC_3+CCFR_4+MEC_2+ATBC+PPS4+USC_HRT_CRC+MCCS_2+DALS_2+PLCO_2+WHI_2+DACHS_1+Colo23+MEC_1+VITAL+DACHS_3+DALS_1+PLCO_1+WHI_1+MECC_1+DACHS_2+HPFS_3_AD+HPFS_1+HPFS_2+NHS_3_AD+NHS_2+NHS_1+PHS+NHS_4+HPFS_4+CLUEII+NHS_5_AD+PLCO_4_AD+EDRN+NCCCSI+NCCCSII+WHI_3+CPSII_2+SMS_AD+HPFS_5_AD+SELECT+PLCO_3+REACH_AD + asp_ref, data = cov, family=binomial)
# 
# 
# results_glm_original <- map_dfr(cov[,67:157], function(x) glm_func_original(x) %>% tidy(), .id = "SNP")
# results_glm_original_filter <- results_glm_original %>% 
#   filter(term == "y") %>% 
#   separate(SNP, into = c("CHR", "BP"), remove = F) %>% 
#   mutate(CHR = as.numeric(CHR), 
#          BP = as.numeric(BP), 
#          P = p.value)

results_glm_original_filter <- readRDS(file = "results_glm_original.rds") %>% 
  filter(term == "y") %>% 
  separate(SNP, into = c("CHR", "BP"), remove = F) %>% 
  mutate(CHR = as.numeric(CHR), 
         BP = as.numeric(BP), 
         P = p.value)

results_G_filter <- results_G %>% 
  filter(SNP %in% results_glm_original_filter$SNP)

x <- inner_join(results_glm_original_filter, results_G_filter[, c('SNP', 'betaG', 'P')], by = 'SNP') %>% 
  mutate(estimate_exp_GLM = exp(estimate),
         estimate_exp_GxEScanR = exp(betaG)) %>% 
  rename(P_GLM = P.x, 
         P_GxEScanR = P.y) %>% 
  dplyr::select(SNP, estimate_exp_GLM, estimate_exp_GxEScanR, P_GLM, P_GxEScanR)

kable(x) %>%
  kable_styling(bootstrap_options = "striped", full_width = F, position = 'left')
```


```{r results_compare2, message=F, echo=T, warning=F, tidy=T, cache=T, include = T}
# lr <- readRDS("results_glm_lrtest.rds")

df <- x %>% 
  mutate(P_GLM = format(P_GLM, digits = 3),
         P_GxEScanR = format(P_GxEScanR, digits = 3))

kable(df) %>%
  kable_styling(bootstrap_options = "striped", full_width = F, position = 'left')
```