---
title: "Mendelian Randomization"
bibliography: /Users/mak/Dropbox/Projects/library.bib
output: 
  #pdf_document:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    theme: "cosmo"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<style type="text/css">
h1 { /* Header 1 */
  font-size: 22px;
  font-weight: bold;
}
h2 { /* Header 2 */
  font-size: 20px;
  font-weight: bold;
}
h3 { /* Header 3 */
  font-size: 18px;
  font-weight: bold;
}
hr {
  border: 1px solid #000;
}
</style>


# Instrument Variable Analysis

## Basic introduction 

### [@Lawlor2008]
endogenous variables - your basic E to associate with outcomes. self-selection = bias. 

intrument variables - related to endogenous variable (first assumption), but unrelated to outcome either directly or indirectly through unobservable variables (second assumption). "In other words, instruments are assumed to be related to the outcome only through the endogenous variable and any observable variables that are included in the analysis as covariates or explanatory variables."

instead of MR, use "instrumental variable analysis with genetic instruments"  

#### How to test second assumption
- distribution of observed covariates by subgroups of the IV (noting that some correlation possible by chance). if there's correlation, stands to reason there might be correlation with unobserved variables too. 
- in genetics, over-identification of IVs is possible. e.g. hasmann and hausman tests. gauge how improved model is with additional IVs jointly modeled. 

choosing suitable endogenous vars - things like smoking, drinking etc are theoretically endogenous because we'd select ourselves in those categories. Things like biomarkers - maybe less self selection (although possible since biomarkers levels could be indicative of behavior etc).. but there's also random effects (measurement instrumentation etc).  
	- in folate MR, need to think about doing functional SNP x Outcome vs IV analyses. 
	
interpretation of IV - for example, a gene that influences smoking initiation vs. addition vs. more intensive smoking. so need to be very explicit about what the IV is measuring in terms of endogenous variables.

might want to stratify IV analysis by relevant variables (akin to interaction of IV by a confounder)
lastly, sensitivity analyses would look at different sets of IVs, different combos. 

<br>

### [@Burgess2015], [@Burgess2015b]

#### Three assumptions:
1. genetic variant is associated with exposure 
2. genetic variant is NOT associated with any confounder of exposure-outcome relationship
3. genetic variant is NOT associated with outcome via any other pathway other than through exposure (no pleiotropy)

IV can be derived from separate dataset, even if they're both meta-analyses. The only important thing is that the two data sets are from the same underlying population  
	- (so i think, even if subset of data, it's ok)

#### Blueprint for analysis
1. identifying dataset for analysis: ideal situation is a single (large) study where you have genetic variants, measured E and D. Alternatively, can have different datasets as sources for IV and IV x D association tests. knowing that assoc is typically weaker between G x D compared to G x E, can also use study derived IV, apply to published G x D data. or use published data for all components why not. 
2. search for IV: a note about biases - weak IV leads to biases towards null. BUT, if same set of individuals used to estimate IV and G x D associations, bias will be in direction of observational association between risk factor and disease/outcome. in subsamples (some overlap), bias depends on extent of overlap -- little overlap = bias towards null
3. validation of IV assumptions: one important thing to do is to test assoc between IV and covariates. easier done with individual data. keep in mind two things - 1) multiple testing - can be too conservative, especially if covars are correlated, 2) even if G is associated with E, doesn't mean it's a bad IV. e.g. G x BMI, G may also be assoc with blood pressure - that's ok because this association may be caused by effect of BMI on BP (vertical pleiotropy). Author recommends two sets of IVs, 'conservative' such that G x Covar P > 0.01, and a 'liberal' more permissive one, such that min p is > bonferroni. 
4. Estimation of the causal effect: 

Testing assumptions:
- Z-X: is it biologically plausible (keep in mind that not all hits are within genetic regions)
- Z-W: can test for each confounder individually, of course can't account for unknown/unmeasured confounder. Can also use GWAS catalogues for common confounders. a few more things - need to consider multiple testing. Also, if exposure is causal to confounder, assumptions are NOT violated. usually - have two sets of covariates, one that's conservative and one that is liberal. 
- population stratification - see @Burgess2015 appendix. 


Question - what if IV is derived from subset of data? I dont' think it should matter



# Examples of MR (preferably in CRC)

## Height and CRC [@Thrift2015]
- Introduction: height (adult) is pos assoc with CRC. Fraught with potential biases etc
- Methods: 
	- IV from GIANT consortium (gene to height). in this analysis - 697 variants associated with height (p<5e-8). Created weighted genetic score predicting height INCREASES, weighted by the per-allele change in height (estimate - increase in cm per on additional risk allele). Then scores are summed for all 697 variants.  
	- Analyses: see paper.. first step - height vs outcome, they did study specific analysis, then meta-analyzed using random effects model. second step - they checked weighted genetic risk score and height to confirm association (checking first assumption... btw what is instrumen bias? see ref 36..). third - check whether genetic risk of height is associated with CRC, colon, and rectal cancer. again, using meta analyses. To check second assumption - linear regression to estimate association of genetically determined height and measured height with a set of potential confounders.. (it sure sounds like they're checking second assumption here)
- Results:  
	- height x outcome: increased risk of CRC. limited to women, null for men
	- IV and height: strong association between genetic risk and height ("(F-statistics: all, ad- justed for sex, partial F-statisticÂ¼1544; women, 916.4; men, 635.5) and the weighted genetic risk score explained 13.3% (partial R2), 13.8% and 12.3% of the variation in height in all persons, women and men, respectively.")
	- IV and CRC/colon/rectal: genetic risk score associated with CRC. Women had stat sig effects. interestingly, more evidence of association among men than with measured height, but not stat sig (would biased height reporting among men also influence IV??). note that this is the weighted genetic risk score, we dont' know in terms of increase in actual height. 
- Discussion: 
	- mechanisms of height x crc associations: steroid/growth hormones = higher cell turnover. as to how, if true, its' for women only = who knows. 	- limitations of IV: assumption tests. 1st assumption - Fstat etc shows it's a strong IV, but only explains 14% of height variation. 2nd assumption - (IV not asociated with confounders). there's assoc IV x meat intake, but after adj for meat intake ests didn't change much. 3rd assumption (no pleiotropy - IV associated with outcome ONLY through endogenous var) - vertical pleiotropy is OK, still operating via bio factors related to height rather than height directly. findings would hold. Horizontal pleiotropy - multiple pathways, not all involving height = bad. to teh best of their knowledge ... 'nah' lol. 


## Application in a two sample mendelian randomization
Two methods available - inverse-variance weighted, and likelihood-based. if precise IV, both give same answer, but otherwise likelihood-based is preferable. 

Recall that in this study, CCFR samples were used for biomarker, which means they overlap with the main study? If that's the case, coefficients Xk (association between genetic and exposure) and Yk (association between genetic and outcome) will be correlated, and need to consider bivariate model of (Xk, Yk) for each genetic variant [@Burgess2013]. 
	(basically, overlapping samples are a headache but not unsurmountable)

# Notes For Analysis
- figure out if samples are overlapping, and if so which


<br><br><br><br>





## read an actual example
### Height x CRC MR
[@Thrift2015]

- GWAS of height found 697 hits, they created weighted genetic risk score by adding all height increasing alleles (or dosage if imputed), weighted by per-allele change in height based on the model. 
- 1) height x crc meta analysis. 
- 2) genetic risk score for height x height in this study population (using controls). again meta-analysis
- 3) genetically determined height x crc . again meta-analysis --> 


### BMI and CRC
[@Thrift2015a]

- Methods: z->y summary stats from another study (GIANT), identified 77 SNPs associated with BMI, constructed weighted genetic risk score - basic addition, weighted by per allele effect on BMI. Stats - need to take close look at confounders. chi2/ttest between confounder and CRC, linear reg/ttest/anove to look at IV, BMI, and confounders.  


<br><br><br><br><br><br><br><br><br><br>
# Bibliography


